### PR TITLE
Add Module `Lesson` methods

### PR DESCRIPTION
**Pull request by shenyih0ng**
## Description

Lesson data are being stored in a nested map with the following hierarchy:

`Semester -> LessonType -> LessonId (e.g. 08J, 1) -> [Lesson] (this is where the timings are stored)` 

Also added a few methods that may be helpful for `pick` and `view`, let me know if you guys need other methods

## TODO

- [x] Add basic unit test to demonstrate usage
- [x] `hasLessonId` method

### CODE REVIEW

File: src/main/java/nus/climods/model/module/LessonType.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+     * @return lesson type
+     */
+    public static LessonType fromValue(String value) {
+        for (LessonType l : LessonType.values()) {
+            if (l.valueStr.equals(value)) {
+                return l;
+            }
+        }
+
+        return LessonType.UNSUPPORTED;
```
Comment (sarrrdin) [inline src/main/java/nus/climods/model/module/LessonType.java:38]: nice defensive coding here!

File: src/main/java/nus/climods/model/module/Module.java [hunk 2]
```diff
@@ -47,6 +60,46 @@ public Module(org.openapitools.client.model.ModuleInformation apiModuleInfo, Str
...
+            SemestersEnum semester = SemestersEnum.fromValue(semesterData.getSemester());
+            List<Lesson> semesterLessons = semesterData.getTimetable();
+            if (semesterLessons == null) {
+                semesterLessons = Collections.emptyList();
+            }
+
+            // Set lesson types for each semester
+            Set<LessonType> semesterLessonsType =
+                semesterLessons.stream().map(Lesson::getLessonType).map(LessonType::fromValue)
+                    .collect(Collectors.toSet());
```
Comment (sarrrdin) [inline src/main/java/nus/climods/model/module/Module.java:83]: nice use of streams!

File: src/main/java/nus/climods/model/module/Module.java [hunk 3]
```diff
@@ -108,6 +161,96 @@ public List<SemestersEnum> getSemesters() {
...
+        requireNonNull(lessonMap);
+        assert lessonMap.containsKey(semester);
+
+        Predicate<Map.Entry<LessonType, ModuleLessonIdMap>> selectableLessonPredicate =
+            lessonEntry -> lessonEntry.getValue().size() > 1;
+
+        return lessonMap.get(semester).entrySet().stream().filter(selectableLessonPredicate).map(Map.Entry::getKey)
+            .collect(Collectors.toSet());
+    }
+
```
Comment (sarrrdin) [inline src/main/java/nus/climods/model/module/Module.java:229]: Actually i was also thinking if we should have a method to return `lessonTypes` that are not selectable? Not sure if we need it here or can just handle inside `PickCommand`.

File: src/test/java/nus/climods/model/module/ModuleTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+import org.openapitools.client.ApiException;
+import org.openapitools.client.api.ModulesApi;
+import org.openapitools.client.model.ModuleInformation;
+import org.openapitools.client.model.SemestersEnum;
+
+class ModuleTest {
+
```
Comment (sarrrdin) [inline src/test/java/nus/climods/model/module/ModuleTest.java:19]: not sure if the negative test cases are required, if there is a need for those we can add them during our v1.4

**Review by sarrrdin — APPROVED** [review]
Just have a few clarification questions, but overall code is clean! LGTM 👍 

TQVM for building the solid infrastructure @shenyih0ng 🙏

File: src/main/java/nus/climods/model/module/Module.java [hunk 5]
```diff
@@ -108,6 +161,96 @@ public List<SemestersEnum> getSemesters() {
...
+    /**
+     * Returns the lessons types that can be selected
+     * <p>
+     * Only lessons that have more than one slot (categorized by the class number) is considered selectable
+     * </p>
+     *
+     * @param semester semester
+     * @return a set of lesson types
+     */
+    public Set<LessonType> getSelectableLessonTypes(SemestersEnum semester) {
```
[the remark this answers is not in the document]
Comment (shenyih0ng) [reply src/main/java/nus/climods/model/module/Module.java:219]: I was thinking whether we want to just interact with `Module` alone for `pick` command. So we will just `Module` to do validation and all the picking, then when we put it into `UserModule` we can just save the lesson details (id, timing etc.)

[the remark this answers is not in the document]
Comment (shenyih0ng) [reply src/main/java/nus/climods/model/module/Module.java:219]: this way we dont need to add all lessons, be it selectable or not, to `UserModule`, instead we can just pick using `Module` and save the relevant info into `UserModule`. What do you think about this?

File: src/main/java/nus/climods/model/module/Module.java (hunk 3 above)
Comment (shenyih0ng) [reply src/main/java/nus/climods/model/module/Module.java:229]: oh yea sure i think its fine, if you need it when implementing `pick` just let me know i can add it

File: src/main/java/nus/climods/model/module/Module.java [hunk 6]
```diff
@@ -108,6 +161,96 @@ public List<SemestersEnum> getSemesters() {
...
+     * @param semester semester
+     * @return lessons
+     */
+    public HashMap<LessonType, ModuleLessonIdMap> getLessons(SemestersEnum semester) {
+        requireNonNull(lessonMap);
+        assert lessonMap.containsKey(semester);
+
+        return lessonMap.get(semester);
+    }
+
```
[the remark this answers is not in the document]
Comment (shenyih0ng) [reply src/main/java/nus/climods/model/module/Module.java:253]: sounds good let me add that in! maybe something like `hasLessonID(String id, SemestersEnum semester, LessonType lesson)`?

File: src/test/java/nus/climods/model/module/ModuleTest.java (hunk 4 above)
Comment (shenyih0ng) [reply src/test/java/nus/climods/model/module/ModuleTest.java:19]: yeap good catch i will add it as a TODO

File: src/main/java/nus/climods/model/module/Module.java (hunk 3 above)
Comment (shenyih0ng) [reply src/main/java/nus/climods/model/module/Module.java:232]: I have added it in the latest commit for convenience

File: src/main/java/nus/climods/model/module/Module.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (shenyih0ng) [reply src/main/java/nus/climods/model/module/Module.java:269]: added in latest commit

File: src/main/java/nus/climods/model/module/Module.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarrrdin) [reply src/main/java/nus/climods/model/module/Module.java:222]: sounds good then we'll focus on only interacting with `Module`

File: src/main/java/nus/climods/model/module/Module.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sarrrdin) [reply src/main/java/nus/climods/model/module/Module.java:269]: wah swee thank you so much

**Review by sarrrdin — APPROVED** [review]
LGTM 👍 Thank you very much @shenyih0ng !

**Pull request merged**